File: 백준/Silver/1213. 팰린드롬 만들기/팰린드롬 만들기.java
import java.util.*;
import java.io.*;

public class Main {

    private static String name;
    private static int[] freq;

    public static void main(String[] args) throws IOException {
        init();
        findAnswer();
    }

    private static void findAnswer() {
        int oddCount = 0;
        char middleChar = '\0';
        for(int i = 0; i < 26; i++) {
            if(freq[i] % 2 == 1) {
                oddCount++;
                middleChar = (char)(i + 'A');
            }
        }
        
        if (oddCount > 1) {
            System.out.println("I'm Sorry Hansoo");
            return;
        }
        
        StringBuilder leftHalf = new StringBuilder();
        for (int i = 0; i < 26; i++) {
            for (int j = 0; j < freq[i] / 2; j++) {
                leftHalf.append((char) (i + 'A'));
            }
        }
        
        StringBuilder palindrome = new StringBuilder(leftHalf);
        if (middleChar != '\0') {
            palindrome.append(middleChar);
        }
        palindrome.append(leftHalf.reverse());

        System.out.println(palindrome);
    }

    private static void init() throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        name = input.readLine();
        freq = new int[26];

        for(char c : name.toCharArray()) {
            freq[c - 'A']++;
        }
    }
}

File: 백준/Silver/1213. 팰린드롬 만들기/팰린드롬 만들기.cc
#include<bits/stdc++.h>
using namespace std;
string input;
int cnt[26];
int main() {
    cin >> input;
    // 앞뒤가 똑같은 문장으로 바꿔야하는데
    // 홀수가 단 하나고 나머지가 짝수거나 모두 짝수여야만 팰린드로믈 만들수 있음
    // 글자가 빠른 순으로 만들어야함
    for (char c : input) {
        cnt[c -'A']++;
    }
    int oc = 0;
    int idx = -1;
    for (int i = 0; i < 26; i++) {
        if (cnt[i] % 2 == 1) {
            oc++; 
            idx = i;
        }
    }

    if (oc >= 2) {
        cout << "I'm Sorry Hansoo";
        exit(0);
    }

    string result;
    for (int i = 0; i < 26; i++) {
        if (cnt[i] >=  2) {
            for (int j = 0; j < cnt[i] / 2; j++) {
                result+='A' + i;
            }
        }
    }
    string temp = result;
    if (idx != -1) {
        result+='A'+idx;
    }
    
    reverse(temp.begin(), temp.end());
    result+=temp;

    cout << result << "\n";

    return 0;
}